File: String/649.Dota2_Senate/C++/649.Dota2_Senate.cpp
class Solution {
public:
    string predictPartyVictory(string senate) 
    {
        //============================//
        // Input type:                //
        //   - string (senate string) //
        // Return type:               //
        //   - string (result string) //
        //============================//

        /*Initialize*/
        ///// Length of senate string /////
        int len_senate = senate.size();

        ///// Record queues (i.e. R, D) /////
        queue<int> record_queue_r, record_queue_d;


        /*Queue-based loop traversal*/
        ///// Step 1: Record indexes with queues /////
        for (int senate_idx = 0; (senate_idx < len_senate); (senate_idx++)) //Whole
        {
            ///// Check if the current indexed-char matched conditions or not /////
            if (senate[senate_idx] == 'R') { record_queue_r.emplace(senate_idx); } //Keep updating, recording
            else { record_queue_d.emplace(senate_idx); } //Keep updating, recording

        } //Whole

        ///// Step 2: Looped-traversal with recorded queues /////
        while ((! record_queue_r.empty()) && (! record_queue_d.empty())) //Part
        {
            ///// Check if the current indexes matched conditions or not /////
            if (record_queue_r.front() < record_queue_d.front())
            {
                record_queue_d.pop(); //Keep updating, popped

                record_queue_r.emplace((record_queue_r.front() + len_senate)); //Keep updating, recording

                record_queue_r.pop(); //Keep updating, popped
            }

            else
            {
                record_queue_r.pop(); //Keep updating, popped

                record_queue_d.emplace((record_queue_d.front() + len_senate)); //Keep updating, recording

                record_queue_d.pop(); //Keep updating, popped
            }

        } //Part

        return (! record_queue_r.empty()) ? "Radiant" : "Dire";
    }
};

File: String/649.Dota2_Senate/Python/649.Dota2_Senate.py
class Solution(object):
    def predictPartyVictory(self, senate):
        """
        :type senate: str
        :rtype: str
        """
        #==============================#
        # Queue-based traversal method #
        #==============================#

        ############
        #Initialize
        ##### Length of senate string #####
        len_senate = len(senate)

        ##### Record queues (i.e. R, D) #####
        record_queue_r, record_queue_d = [], []


        ############################
        #Queue-based loop traversal
        ##### Step 1: Record indexes with queues #####
        for senate_idx in range(len_senate):

            ##### Check if the current indexed-char matched conditions or not #####
            if (senate[senate_idx] == 'R'):
                record_queue_r.append(senate_idx) #Keep updating/recording
            else:
                record_queue_d.append(senate_idx) #Keep updating/recording

        ##### Step 2: Looped-traversal with recorded queues #####
        while (record_queue_r and record_queue_d):

            ##### Check if the current index matched conditions or not ##### 
            if (record_queue_r[0] < record_queue_d[0]):
                record_queue_d.pop(0) #Keep updating/popped

                record_queue_r.append((record_queue_r[0] + len_senate)) #Keep updating/recording

                record_queue_r.pop(0) #Keep updating/popped
            
            else:
                record_queue_r.pop(0) #Keep updating/popped

                record_queue_d.append((record_queue_d[0] + len_senate)) #Keep updating/recording

                record_queue_d.pop(0) #Keep updating/popped
        
        return ("Radiant" if (record_queue_r) else "Dire")
